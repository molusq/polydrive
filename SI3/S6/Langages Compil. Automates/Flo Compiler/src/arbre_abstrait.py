import enum
import json
from typing import List, Optional, TypeAlias

JSON: TypeAlias = dict[str,
                       "JSON"] | list["JSON"] | str | int | float | bool | None


class Type(enum.Enum):
    INCONNU = enum.auto()
    ENTIER = enum.auto()
    BOOLEEN = enum.auto()
    FONCTION = enum.auto()
    VIDE = enum.auto()

    @staticmethod
    def from_str(s: str) -> "Type":
        if s == "entier":
            return Type.ENTIER
        if s == "booleen":
            return Type.BOOLEEN
        return Type.INCONNU


class AST:

    def to_json(self) -> JSON:
        print(self.__class__.__name__)
        raise NotImplementedError("to_json not implemented")

    def type(self) -> Type:
        return Type.INCONNU

    def __str__(self):
        return json.dumps(self.to_json(), indent=2)


class Programme(AST):

    def __init__(self, instructions: List[AST]):
        self.instructions = instructions

    def type(self) -> Type:
        raise NotImplementedError("Programme has no type")
        return Type.AUTRE

    def to_json(self) -> JSON:
        return {"instructions": [i.to_json() for i in self.instructions]}

    def addInstruction(self, instruction: AST):
        self.instructions.append(instruction)


class Operation(AST):

    def __init__(self, op: str, exp1: AST, exp2: AST):
        self.lhs = exp1
        self.op = op
        self.rhs = exp2

    def type(self) -> Type:
        if self.lhs.type() == self.rhs.type():
            if self.op in ["<", "<=", ">", ">=", "==", "!="]:
                return Type.BOOLEEN
            return self.lhs.type()
        print("op:", self.lhs.to_json(), self.rhs.to_json())
        return Type.INCONNU

    def to_json(self) -> JSON:
        return {
            "op": self.op,
            "lhs": self.lhs.to_json(),
            "rhs": self.rhs.to_json()
        }


class OperationUnaire(AST):

    def __init__(self, op: str, exp: AST):
        self.op = op
        self.exp = exp

    def type(self) -> Type:
        return self.exp.type()

    def to_json(self) -> JSON:
        return {"op": self.op, "exp": self.exp.to_json()}


class Entier(AST):

    def __init__(self, valeur: int):
        self.valeur = valeur

    def type(self) -> Type:
        return Type.ENTIER

    def __eq__(self, rhs: object) -> bool:
        if isinstance(rhs, Entier):
            return self.valeur == rhs.valeur
        return self.valeur == rhs

    def to_json(self) -> JSON:
        return {"entier": self.valeur}


class Booleen(AST):

    def __init__(self, valeur: bool):
        self.valeur = valeur

    def type(self) -> Type:
        return Type.BOOLEEN

    def __eq__(self, rhs: object) -> bool:
        if isinstance(rhs, Booleen):
            return self.valeur == rhs.valeur
        return self.valeur == rhs

    def to_json(self) -> JSON:
        return {"booleen": self.valeur}


class Identifiant(AST):

    def __init__(self, valeur: str):
        self.valeur = valeur

    def type(self) -> Type:
        return Type.AUTRE

    def __eq__(self, rhs: object) -> bool:
        if isinstance(rhs, Identifiant):
            return self.valeur == rhs.valeur
        return self.valeur == rhs

    def to_json(self) -> JSON:
        return {"identifiant": self.valeur}


class If(AST):

    def __init__(self, condition: AST, body: Programme,
                 orelse: Optional[Programme]):
        self.condition = condition
        self.body = body
        self.orelse = orelse or Programme([])

    def type(self) -> Type:
        return Type.AUTRE

    def add_elif(self, condition: AST, body: Programme):
        if self.orelse.instructions and isinstance(
                self.orelse.instructions[-1], If):
            self.orelse.instructions[-1].add_elif(condition, body)
        else:
            self.orelse.addInstruction(If(condition, body, None))

    def add_else(self, body: Programme):
        self.add_elif(Booleen(True), body)

    def to_json(self) -> JSON:
        return {
            "condition": self.condition.to_json(),
            "body": self.body.to_json(),
            "orelse": self.orelse.to_json()
        }


class While(AST):

    def __init__(self, condition: AST, body: Programme):
        self.condition = condition
        self.body = body

    def type(self) -> Type:
        return Type.AUTRE

    def to_json(self) -> JSON:
        return {
            "condition": self.condition.to_json(),
            "body": self.body.to_json()
        }


class Declaration(AST):

    def __init__(self, name: Identifiant, _type: Type, value: Optional[AST]):
        self.name = name
        self._type = _type
        self.value = value

    def type(self) -> Type:
        return self._type

    def to_json(self) -> JSON:
        return {
            "name": self.name.to_json(),
            "type": self.type().name,
            "value": self.value.to_json() if self.value else None
        }


class Assignment(AST):

    def __init__(self, name: Identifiant, value: AST):
        self.name = name
        self.value = value

    def type(self) -> Type:
        return Type.AUTRE

    def to_json(self) -> JSON:
        return {"name": self.name.to_json(), "value": self.value.to_json()}


class Return(AST):

    def __init__(self, value: AST):
        self.value = value

    def type(self) -> Type:
        return Type.AUTRE

    def to_json(self) -> JSON:
        return {"return_value": self.value.to_json()}


class Function(AST):

    def __init__(self, name: Identifiant, args: List[Declaration],
                 body: Programme, return_type: Type):
        self.name = name
        self.args = args
        self.body = body
        self.return_type = return_type
        self.stack_size = 0

    def type(self) -> Type:
        return Type.FONCTION

    def to_json(self) -> JSON:
        return {
            "name": self.name.to_json(),
            "args": [a.to_json() for a in self.args],
            "body": self.body.to_json(),
            "return_type": self.return_type.name,
            "stack_size": self.stack_size
        }


class FunctionArgs(AST):

    def __init__(self, args: List[AST]):
        self.args = args

    def type(self) -> Type:
        return Type.AUTRE

    def to_json(self) -> JSON:
        return list(map(lambda x: x.to_json(), self.args))

    def __getitem__(self, i: int):
        return self.args[i]

    def __iter__(self):
        return iter(self.args)

    def append(self, arg: AST):
        self.args.append(arg)

    def __len__(self):
        return len(self.args)


class AppelFonction(AST):

    def __init__(self, name: Identifiant, args: FunctionArgs):
        self.name = name
        self.args = args

    def type(self) -> Type:
        return Type.INCONNU

    def to_json(self) -> JSON:
        return {"function_name": self.name.valeur, "args": self.args.to_json()}
