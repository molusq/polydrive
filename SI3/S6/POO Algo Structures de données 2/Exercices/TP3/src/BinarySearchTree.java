package ads.poo2.lab3.bst2.ORIGIN;

import java.util.*;

/**
 * A class for Binary Search Trees
 */
public class BinarySearchTree<T extends Comparable<? super T>> implements Iterable<T> {

    private T data;
    private BinarySearchTree<T> left;
    private BinarySearchTree<T> right;

    //return the value of the root of the tree
    public T getElement() {
        return data;
    }


    protected BinarySearchTree<T> getLeft() {
        return left;
    }

    protected void setLeft(BinarySearchTree<T> left) {
        this.left = left;
    }

    protected BinarySearchTree<T> getRight() {
        return right;
    }

    protected void setRight(BinarySearchTree<T> right) {
        this.right = right;
    }



    /**
     * Construct the tree.
     */
    public BinarySearchTree(T data, BinarySearchTree<T> left, BinarySearchTree<T> right) {
        this.data = data;
        this.left = left;
        this.right = right;
    }

    public BinarySearchTree(T data) {
        this(data, null, null);
    }

    public BinarySearchTree() {
        this(null, null, null);
    }

    /////////////// isEmpty

    /**
     * Test if the tree is logically empty.
     *
     * @return true if empty, false otherwise.
     */
    public boolean isEmpty() {
        if(this.data == null && this.left == null && this.right == null)
            return true;
        return false;
    }

    /////////////// makeEmpty  

    /**
     * Make the tree logically empty.
     */
    public void makeEmpty() {
        this.data = null;
        this.left = null;
        this.right = null;
    }
    /////////////// contains

    /**
     * Find an item in the tree.
     *
     * @param x the item to search for.
     * @return true if not found.
     */
    public boolean contains(T x) {
        if (x==null ||isEmpty()){
            return false;
        }
        int compareResult = x.compareTo(this.getElement());
        if(compareResult ==0){
            return true;
        }if(compareResult < 0 && left !=null){
            return left.contains(x);
        }else if(compareResult >0 && right != null){
            return right.contains(x);
        }else{
            return false;
        }
    }

    //////////////// size ////////////////
    // The size of a BN is its number of
    // non-null nodes
    public int getSize() {

        if (isEmpty()) {
            return 0;
        }
        if (isLeaf()) {
            return 1;
        }
        if (left == null) {
            return 1+right.getSize();
        }
        else if (right == null) {
            return 1+left.getSize();
        }
        else {
            return 1+left.getSize()+right.getSize();
        }

    }

    public boolean isLeaf() {
        //to do
        return left == null && right == null;
    }

    /////////////// insert

    /**
     * Insert into the tree; duplicates are ignored.
     *
     * @param x the item to insert.
     */
    public void insert(T x) {
        if (x==null ||isEmpty()){
            this.data = x;
        }
        else{
            int compareResult = x.compareTo(this.getElement());
            if(compareResult < 0 ){
                if(left == null){
                    left = new BinarySearchTree<>();
                }
                left.insert(x);
            }else if(compareResult >0 ){
                if(right == null){
                    right = new BinarySearchTree<>();
                }
                right.insert(x);
            }
        }
    }

    public T findMin() {
        if(this.left == null){
            return this.getElement();
        }
        else{
            return left.findMin();
        }
    }

    /////////////// findMax

    /**
     * Find the largest item in the tree.
     *
     * @return the largest item or null if empty.
     */
    public T findMax() {
        if(this.right == null){
            return this.getElement();
        }
        else{
            return right.findMin();
        }
    }

    /////////////// remove

    /**
     * Remove from the tree. Nothing is done if x is not found.
     * In the test, we replace the removed element by the largest element of the left subtree
     * @param x the item to remove.
     */
    public void remove(T x) {
        int compareResult = x.compareTo(this.getElement());
        if(compareResult ==0){
            if (isLeaf()) {
                makeEmpty();
                return;
            }
        }else if(compareResult < 0 && left !=null){
            left.remove(x);
        }else if(compareResult >0 && right != null){
            right.remove(x);
        }
    }

    /////////////// removeLessThan



    /**ree all the elements
     * less than min
     * Remove from the t
     *
     * @param min the minimum value left in the tree
     */
    public void removeLessThan(T min) {
        root = removeLessThan( root, min );
    }
    
    private BinaryNode<AnyType> removeLessThan(BinaryNode<AnyType> t, AnyType min) {
        if ( t == null )
            return null;
        if ( t.element.compareTo(min) < 0 )
            return removeLessThan(t.right,min);
        t.left = removeLessThan(t.left,min);
        return t;
    }



    /////////////// removeGreaterThana

    /**
     * Remove from the tree all the elements
     * greater than max
     *
     * @param max the maximum value left in the tree
     */
    public void removeGreaterThan(T max) {

    }

    /////////////// toSortedList

    /**
     * Return a sorted list (increasing) of all
     * the elements of the tree
     *
     * @return the sorted list of all the elements of the tree
     */
    public List<T> toSortedList() {
        return null;
    }

    /////////////// sorted list to binary search tree

    /**
     * Build a binary search tree with all the
     * elements of the list
     *
     * @param list a sorted (increasing) list of elements
     */
    public BinarySearchTree(List<T> list) {
        makeTree(list, 0, list.size() - 1);
    }

    //Usefull method to build a binary search tree  from a sorted list
    //The list is divided in two parts, the first part is used to build
    //the left subtree, the second part is used to build the right subtree
    private BinarySearchTree<T> makeTree(List<T> list, int i, int j) {
        return null;
    }


    /////////////// iterator on binary search tree


    /**
     * Return an iterator over the elements of the tree.
     * The elements are enumerated in increasing order.
     */
    public Iterator<T> iterator() {
        return new BinarySearchTree<T>.BSTiterator(this);
    }


    /**
     * Inner class to build iterator over the elements of
     * a tree
     */
    private class BSTiterator implements Iterator<T> {

        // we must push some binary nodes on the stack

        Deque<BinarySearchTree<T>> stack;


        /**
         * Build an iterator over the binary node n.
         * The elements are enumerated in increasing order.
         */
        BSTiterator(BinarySearchTree<T> n) {
            stack = new ArrayDeque<>();
            // push all the left nodes on the stack
            //TODO
        }

        /**
         * Check if there are more elements in the
         * iterator
         */
        public boolean hasNext() {
            return false;
        }

        /**
         * Return and remove the next element from
         * the iterator
         */
        public T next() {
            return null;
        }

        /**
         * Unsupported operation
         */
        @Override
        public void remove() {
            throw new UnsupportedOperationException("remove");
        }
    }

    ////////////////////////////////////////////////////
    // Convenience method to print a tree
    ////////////////////////////////////////////////////


    public void display() {
        display(this, "", "");
    }
    private void display(BinarySearchTree<T> t, String r, String p) {
        if (t == null || t.isEmpty()) {
            System.out.println(r);
        } else {
            String rs = t.getElement().toString();
            System.out.println(r + rs);
            if (t.getLeft() != null || t.getRight() != null) {
                String rr = p + '|' + makeString('_', rs.length()) + ' ';
                display(t.getLeft(), rr, p + '|' + makeString(' ', rs.length() + 1));
                System.out.println(p + '|');
                display(t.getRight(), rr, p + makeString(' ', rs.length() + 2));
            }
        }
    }

    private String makeString(char c, int k) {
        return String.valueOf(c).repeat(Math.max(0, k));
    }


    ////////////////////////////////////////////////////
    // Convenience methods to build a list of integer from a string
    ////////////////////////////////////////////////////

    private static List<Integer> read(String inputString) {
        List<Integer> list = new LinkedList<>();
        Scanner input = new Scanner(inputString);
        while (input.hasNextInt())
            list.add(input.nextInt());
        input.close();
        return list;
    }

    /**
     * A short main for quick testing
     */
    public static void main(String[] args)  {
            List<Integer> list = read("50 30 70 20 40 80 60");
            BinarySearchTree<Integer> bst = new BinarySearchTree<>();
            for (Integer n : list)
                bst.insert(n);
            bst.display();
            //From the lesson
            list = Arrays.asList(12, 17, 21, 19, 14, 13, 16, 9, 11, 10, 5, 8);
            System.out.println("---- From the lesson one by one");
            bst = new BinarySearchTree<>();
            //To control the position
            for (Integer n : list) {
                bst.insert(n);
            }
            System.out.println("Before insert 3");
            bst.display();
            bst.insert(3);
            System.out.println("After insert 3");
            bst.display();
            bst.remove(17);
            System.out.println("After remove 17");
            bst.display();
            System.out.println("---- From the lesson using initialisation with Arrays.asList");
            bst = new BinarySearchTree<>(list);
            bst.display();
            bst.insert(3);
            bst.display();
            bst.remove(17);
            bst.display();
    }
}
