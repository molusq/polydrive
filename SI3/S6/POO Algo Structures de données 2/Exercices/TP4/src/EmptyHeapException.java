
/**
 * A class exception for empty heap
 */
public class EmptyHeapException extends Exception {
	
	public EmptyHeapException() {
		super();
	}

}
