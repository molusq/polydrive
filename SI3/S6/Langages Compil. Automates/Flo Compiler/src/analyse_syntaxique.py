# type: ignore

from sly import Parser
from src.analyse_lexicale import FloLexer
import src.arbre_abstrait as arbre_abstrait


class FloParser(Parser):

    def error(self, token):
        '''
        Default error handling function.  This may be subclassed.
        '''
        if token:
            lineno = getattr(token, 'lineno', 0)
            if lineno:
                print(
                    f'sly: Syntax error at line {lineno}, token={token.type}\n'
                )
            else:
                print(f'sly: Syntax error, token={token.type}')
        else:
            print('sly: Parse error in input. EOF\n')
        print(list(self.symstack))

    # On récupère la liste des lexèmes de l'analyse lexicale
    tokens = FloLexer.tokens

    precedence = (
        ('left', 'ET'),
        ('left', 'OU'),
        ('right', 'NON'),
        ('right', 'EGAL'),
        ('left', 'DIFFERENT', 'INFERIEUR_OU_EGAL', 'SUPERIEUR_OU_EGAL',
         'INFERIEUR', 'SUPERIEUR'),
        ('left', '+', '-'),
        ('left', '*', '/', '%'),
        ('nonassoc', 'UMOINS'),
    )

    # Règles gramaticales et actions associées
    @_('instruction')
    def prog(self, p):
        return arbre_abstrait.Programme([p.instruction])

    @_('prog instruction')
    def prog(self, p):
        p.prog.addInstruction(p.instruction)
        return p.prog

    @_('expr ";"')
    def instruction(self, p):
        return p.expr

    @_('bool')
    def expr(self, p):
        return p[0]

    @_('BOOLEEN')
    def bool(self, p):
        return arbre_abstrait.Booleen(p[0])

    @_('bool ET bool')
    def bool(self, p):
        return arbre_abstrait.Operation('et', p[0], p[2])

    @_('bool OU bool')
    def bool(self, p):
        return arbre_abstrait.Operation('ou', p[0], p[2])

    @_('NON bool')
    def bool(self, p):
        return arbre_abstrait.OperationUnaire('non', p[1])

    @_('exprArith')
    def bool(self, p):
        return p[0]

    @_('exprArith "+" exprArith', 'exprArith "-" exprArith',
       'exprArith "*" exprArith', 'exprArith "/" exprArith',
       'exprArith "%" exprArith', 'exprArith INFERIEUR exprArith',
       'exprArith SUPERIEUR exprArith', 'exprArith EGAL exprArith',
       'exprArith DIFFERENT exprArith',
       'exprArith INFERIEUR_OU_EGAL exprArith',
       'exprArith SUPERIEUR_OU_EGAL exprArith')
    def exprArith(self, p):
        return arbre_abstrait.Operation(p[1], p[0], p[2])

    @_('"-" exprArith %prec UMOINS')
    def exprArith(self, p):
        return arbre_abstrait.OperationUnaire('-', p[1])

    @_('"(" bool ")"')
    def exprArith(self, p):
        return p[1]

    @_('facteur')
    def exprArith(self, p):
        return p[0]

    @_('identifiant "(" function_arg ")"')
    def facteur(self, p):
        return arbre_abstrait.AppelFonction(p.identifiant, p.function_arg)

    @_('identifiant "("  ")"')
    def facteur(self, p):
        return arbre_abstrait.AppelFonction(p.identifiant,
                                            arbre_abstrait.FunctionArgs([]))

    @_('ENTIER')
    def facteur(self, p):
        return arbre_abstrait.Entier(p.ENTIER)

    @_('IDENTIFIANT')
    def facteur(self, p):
        return arbre_abstrait.Identifiant(p.IDENTIFIANT)

    @_('IDENTIFIANT')
    def identifiant(self, p):
        return arbre_abstrait.Identifiant(p.IDENTIFIANT)

    @_('SI "(" expr ")" "{" prog "}"')
    def si(self, p):
        return arbre_abstrait.If(p.expr, p.prog, None)

    @_('si SINON SI "(" expr ")" "{" prog "}"')
    def si(self, p):
        p.si.add_elif(p.expr, p.prog)
        return p.si

    @_('si')
    def cond(self, p):
        return p.si

    @_('si SINON "{" prog "}"')
    def cond(self, p):
        p.si.add_else(p.prog)
        return p.si

    @_('TANTQUE "(" expr ")" "{" prog "}"')
    def instruction(self, p):
        return arbre_abstrait.While(p.expr, p.prog)

    @_('cond')
    def instruction(self, p):
        return p.cond

    @_('TYPE identifiant')
    def decl(self, p):
        return arbre_abstrait.Declaration(p.identifiant,
                                          arbre_abstrait.Type.from_str(p.TYPE),
                                          None)

    @_('decl ";"')
    def instruction(self, p):
        return p.decl

    @_('TYPE identifiant "=" expr ";"')
    def instruction(self, p):
        return arbre_abstrait.Declaration(p.identifiant,
                                          arbre_abstrait.Type.from_str(p.TYPE),
                                          p.expr)

    @_('identifiant "=" expr ";"')
    def instruction(self, p):
        return arbre_abstrait.Assignment(p.identifiant, p.expr)

    @_('RETOURNER expr ";"')
    def instruction(self, p):
        return arbre_abstrait.Return(p.expr)

    @_('function_arg "," expr')
    def function_arg(self, p):
        p.function_arg.append(p.expr)
        return p.function_arg

    @_('expr')
    def function_arg(self, p):
        return arbre_abstrait.FunctionArgs([p.expr])

    @_('decl')
    def function_decl_arg(self, p):
        return [p.decl]

    @_('function_decl_arg "," decl')
    def function_decl_arg(self, p):
        p.function_decl_arg.append(p.decl)
        return p.function_decl_arg

    @_('TYPE identifiant "(" function_decl_arg ")" "{" prog "}"')
    def function(self, p):
        return arbre_abstrait.Function(p.identifiant, p.function_decl_arg,
                                       p.prog,
                                       arbre_abstrait.Type.from_str(p.TYPE))

    @_('TYPE identifiant "(" ")" "{" prog "}"')
    def function(self, p):
        return arbre_abstrait.Function(p.identifiant, [], p.prog,
                                       arbre_abstrait.Type.from_str(p.TYPE))

    @_('function')
    def instruction(self, p):
        return p.function


def analyse_syntaxique(input: str) -> arbre_abstrait.Programme:
    lexer = FloLexer()
    parser = FloParser()
    return parser.parse(lexer.tokenize(input))
