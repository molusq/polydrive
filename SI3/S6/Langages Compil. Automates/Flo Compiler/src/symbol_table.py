from typing import Dict, List

from src.arbre_abstrait import Declaration, Identifiant, Programme, Type, Function


class Variable:

    def __init__(self, name: Identifiant, type: Type, offset: int, depth: int):
        self.name = name
        self.type = type
        self.offset = offset
        self.depth = depth


Symbol = Function | Variable


class SymbolTable:

    def __init__(self):
        self._symbols: Dict[str, Symbol] = {}

    @staticmethod
    def from_program(program: Programme):
        symbol_table = SymbolTable()
        for instruction in program.instructions:
            if isinstance(instruction, Function):
                symbol_table.add(instruction)
        return symbol_table

    def check_presence(self, sym: Identifiant):
        if not sym in self:
            raise Exception(f"Le symbole {sym} n'est pas défini")

    def add(self, sym: Symbol):
        if sym in self:
            raise Exception(f"Le symbole {sym.name} est déjà défini")
        self._symbols[sym.name.valeur] = sym

    def get(self, sym: Identifiant) -> Symbol:
        self.check_presence(sym)
        return self._symbols[sym.valeur]

    def remove(self, sym: Identifiant):
        self.check_presence(sym)
        del self._symbols[sym.valeur]

    def type(self, func: Identifiant) -> Type:
        self.check_presence(func)
        return self._symbols[func.valeur].return_type

    def args(self, func: Identifiant) -> List[Declaration]:
        self.check_presence(func)
        return self._symbols[func.valeur].args

    def memory_size(self, sym: Identifiant) -> int:
        self.check_presence(sym)
        func = self.get(sym)
        return 4 * len(func.args) + func.stack_size

    def memory_size_locals(self, sym: Identifiant) -> int:
        self.check_presence(sym)
        func = self.get(sym)
        return func.stack_size

    def next_address(self) -> int:
        return max([
            sym.offset
            for sym in self._symbols.values() if type(sym) == Variable
        ],
                   default=0) + 4

    def end_block(self, depth: int):
        in_depth = {
            k: v
            for k, v in self._symbols.items()
            if type(v) == Variable and v.depth == depth
        }
        for k in in_depth:
            del self._symbols[k]
        return in_depth

    def __contains__(self, sym: Symbol | Identifiant):
        if isinstance(sym, Identifiant):
            return sym.valeur in self._symbols

        return sym.name.valeur in self._symbols

    def __str__(self):
        return '\n'.join([str(symbol) for symbol in self._symbols.values()])