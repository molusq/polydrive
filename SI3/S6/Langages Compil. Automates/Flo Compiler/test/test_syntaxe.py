import syrupy
from src.analyse_syntaxique import analyse_syntaxique
import json


def get_out(input) -> str:
    out = analyse_syntaxique(input).to_json()
    return json.dumps(out, indent=2)


def test_syntaxe(snapshot):
    input = """
    ecrire(- (a + b - c) * d / e % f);
    fonction1(a et non(b ou c));
    bonsoir(a == b != c < d <= e > f >= g);
    """

    assert get_out(input) == snapshot


def test_parse_while(snapshot):
    input = """
   tantque (a) {
       ecrire(a);
   }
   """

    assert get_out(input) == snapshot


def test_parse_if(snapshot):
    input = """
    si (a) {
        ecrire(a);
    }
    sinon si (b) {
        ecrire(b);
    }
    sinon si (c) {
        ecrire(c);
        ecrire(d);
    }
    sinon {
        ecrire(c);
    }
    """

    assert get_out(input) == snapshot


def test_invalid_parse_if():
    INVALID = [
        """
    si (a) {
        ecrire(a);
    }
    sinon {
        ecrire(d);
    }
    sinon si (b) {
        ecrire(b);
    }
    """,
        """
    sinon {
        ecrire(a);
    }
    """,
    ]

    for input in INVALID:
        try:
            get_out(input)
            assert False
        except Exception:
            pass


def test_declare_assign(snapshot):
    input = """
    entier a;
    a = 1;
    booleen b = vrai;
    """

    assert get_out(input) == snapshot


def test_return(snapshot):
    input = """
    retourner 1;
    """

    assert get_out(input) == snapshot


def test_function_call(snapshot):
    input = """
    fonction1(a, b, c);
    """

    assert get_out(input) == snapshot


def test_function_declare(snapshot):
    input = """
    entier fonction1(entier a, booleen b, booleen c) {
        ecrire(a + b % c);
        retourner a;
    }

    booleen fonction2 () {
        retourner 1;
    }

    fonction1(a, b, c);
    fonction2();
    """

    assert get_out(input) == snapshot
