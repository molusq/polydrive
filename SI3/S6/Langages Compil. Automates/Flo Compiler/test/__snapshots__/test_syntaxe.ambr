# serializer version: 1
# name: test_declare_assign
  '''
  {
    "instructions": [
      {
        "name": {
          "identifiant": "a"
        },
        "type": "ENTIER",
        "value": null
      },
      {
        "name": {
          "identifiant": "a"
        },
        "value": {
          "entier": 1
        }
      },
      {
        "name": {
          "identifiant": "b"
        },
        "type": "BOOLEEN",
        "value": {
          "identifiant": "vrai"
        }
      }
    ]
  }
  '''
# ---
# name: test_function_call
  '''
  {
    "instructions": [
      {
        "function_name": "fonction1",
        "args": [
          {
            "identifiant": "a"
          },
          {
            "identifiant": "b"
          },
          {
            "identifiant": "c"
          }
        ]
      }
    ]
  }
  '''
# ---
# name: test_function_declare
  '''
  {
    "instructions": [
      {
        "name": {
          "identifiant": "fonction1"
        },
        "args": [
          {
            "name": {
              "identifiant": "a"
            },
            "type": "ENTIER",
            "value": null
          },
          {
            "name": {
              "identifiant": "b"
            },
            "type": "BOOLEEN",
            "value": null
          },
          {
            "name": {
              "identifiant": "c"
            },
            "type": "BOOLEEN",
            "value": null
          }
        ],
        "body": {
          "instructions": [
            {
              "function_name": "ecrire",
              "args": [
                {
                  "op": "+",
                  "lhs": {
                    "identifiant": "a"
                  },
                  "rhs": {
                    "op": "%",
                    "lhs": {
                      "identifiant": "b"
                    },
                    "rhs": {
                      "identifiant": "c"
                    }
                  }
                }
              ]
            },
            {
              "return_value": {
                "identifiant": "a"
              }
            }
          ]
        },
        "return_type": "ENTIER",
        "stack_size": 0
      },
      {
        "name": {
          "identifiant": "fonction2"
        },
        "args": [],
        "body": {
          "instructions": [
            {
              "return_value": {
                "entier": 1
              }
            }
          ]
        },
        "return_type": "BOOLEEN",
        "stack_size": 0
      },
      {
        "function_name": "fonction1",
        "args": [
          {
            "identifiant": "a"
          },
          {
            "identifiant": "b"
          },
          {
            "identifiant": "c"
          }
        ]
      },
      {
        "function_name": "fonction2",
        "args": []
      }
    ]
  }
  '''
# ---
# name: test_parse_if
  '''
  {
    "instructions": [
      {
        "condition": {
          "identifiant": "a"
        },
        "body": {
          "instructions": [
            {
              "function_name": "ecrire",
              "args": [
                {
                  "identifiant": "a"
                }
              ]
            }
          ]
        },
        "orelse": {
          "instructions": [
            {
              "condition": {
                "identifiant": "b"
              },
              "body": {
                "instructions": [
                  {
                    "function_name": "ecrire",
                    "args": [
                      {
                        "identifiant": "b"
                      }
                    ]
                  }
                ]
              },
              "orelse": {
                "instructions": [
                  {
                    "condition": {
                      "identifiant": "c"
                    },
                    "body": {
                      "instructions": [
                        {
                          "function_name": "ecrire",
                          "args": [
                            {
                              "identifiant": "c"
                            }
                          ]
                        },
                        {
                          "function_name": "ecrire",
                          "args": [
                            {
                              "identifiant": "d"
                            }
                          ]
                        }
                      ]
                    },
                    "orelse": {
                      "instructions": []
                    }
                  }
                ]
              }
            },
            {
              "function_name": "ecrire",
              "args": [
                {
                  "identifiant": "c"
                }
              ]
            }
          ]
        }
      }
    ]
  }
  '''
# ---
# name: test_parse_while
  '''
  {
    "instructions": [
      {
        "condition": {
          "identifiant": "a"
        },
        "body": {
          "instructions": [
            {
              "function_name": "ecrire",
              "args": [
                {
                  "identifiant": "a"
                }
              ]
            }
          ]
        }
      }
    ]
  }
  '''
# ---
# name: test_return
  '''
  {
    "instructions": [
      {
        "return_value": {
          "entier": 1
        }
      }
    ]
  }
  '''
# ---
# name: test_syntaxe
  '''
  {
    "instructions": [
      {
        "function_name": "ecrire",
        "args": [
          {
            "op": "%",
            "lhs": {
              "op": "/",
              "lhs": {
                "op": "*",
                "lhs": {
                  "op": "-",
                  "exp": {
                    "op": "-",
                    "lhs": {
                      "op": "+",
                      "lhs": {
                        "identifiant": "a"
                      },
                      "rhs": {
                        "identifiant": "b"
                      }
                    },
                    "rhs": {
                      "identifiant": "c"
                    }
                  }
                },
                "rhs": {
                  "identifiant": "d"
                }
              },
              "rhs": {
                "identifiant": "e"
              }
            },
            "rhs": {
              "identifiant": "f"
            }
          }
        ]
      },
      {
        "function_name": "fonction1",
        "args": [
          {
            "op": "et",
            "lhs": {
              "identifiant": "a"
            },
            "rhs": {
              "op": "non",
              "exp": {
                "op": "ou",
                "lhs": {
                  "identifiant": "b"
                },
                "rhs": {
                  "identifiant": "c"
                }
              }
            }
          }
        ]
      },
      {
        "function_name": "bonsoir",
        "args": [
          {
            "op": "==",
            "lhs": {
              "identifiant": "a"
            },
            "rhs": {
              "op": ">=",
              "lhs": {
                "op": ">",
                "lhs": {
                  "op": "<=",
                  "lhs": {
                    "op": "<",
                    "lhs": {
                      "op": "!=",
                      "lhs": {
                        "identifiant": "b"
                      },
                      "rhs": {
                        "identifiant": "c"
                      }
                    },
                    "rhs": {
                      "identifiant": "d"
                    }
                  },
                  "rhs": {
                    "identifiant": "e"
                  }
                },
                "rhs": {
                  "identifiant": "f"
                }
              },
              "rhs": {
                "identifiant": "g"
              }
            }
          }
        ]
      }
    ]
  }
  '''
# ---
